### PR TITLE
Feat GCCJITlibs

### PR DESCRIPTION
**Pull request by JoGei**
# 1. GCC support for extentending JITlibs

- added feature for GCCJIT to support JIT linking of extending JIT libraries, e.g. softfloat for FPU.

# 2. Remote fetch of softfloat library and local build

- removed pre-compiled softfloat sources
- combined old pre-compiled softfloat lib and locally built dbtrise_fp_funcs library to one new ~static~ shared softfloat lib. 
-> Size reduction of linked libraries for JIT.

### CODE REVIEW

File: src/jitlibs/softfloat/CMakeLists.txt [hunk 1]
```diff
@@ -1,14 +1,405 @@
...
+	file(COPY ${softfloat_srcs_SOURCE_DIR}/source/include/softfloat.h DESTINATION ${softfloat_srcs_SOURCE_DIR})
+	file(RENAME ${softfloat_srcs_SOURCE_DIR}/softfloat.h ${softfloat_srcs_SOURCE_DIR}/softfloat_orig.h )
+	FetchContent_GetProperties(softfloat_srcs)
+	set(EXTERNAL_INSTALL_LOCATION ${softfloat_srcs_SOURCE_DIR})
+
+##################### SOFTFLOAT UTILITY PROJECT (berkeley softfloat) #####################
+
+EXECUTE_PROCESS( COMMAND uname -m COMMAND tr -d '\n' OUTPUT_VARIABLE ARCHITECTURE )
+IF(WIN32)
+	IF(NOT CMAKE_CL_64)
```
Comment (rafzi) [inline src/jitlibs/softfloat/CMakeLists.txt:20]: have you tested this? it seems like it wouldn't work, because when using mingw, you are not using msvc cl. The documentation proposes the proper alternative.

File: src/jitlibs/softfloat/CMakeLists.txt [hunk 2]
```diff
@@ -1,14 +1,405 @@
...
+	${EXTERNAL_INSTALL_LOCATION}/build/${PLATFORM_N}/platform.h
+	${EXTERNAL_INSTALL_LOCATION}/${INC}/primitiveTypes.h
+	${EXTERNAL_INSTALL_LOCATION}/${INC}/primitives.h
+	${EXTERNAL_INSTALL_LOCATION}/${INC}/softfloat_types.h
+	${EXTERNAL_INSTALL_LOCATION}/${INC}/internals.h
+	${EXTERNAL_INSTALL_LOCATION}/source/${SPECIALIZE}/specialize.h
+	${EXTERNAL_INSTALL_LOCATION}/${INC}/softfloat.h
+)
+
+SET(CMAKE_C_ARCHIVE_CREATE "<CMAKE_AR> crs <TARGET> <LINK_FLAGS> <OBJECTS>")
```
Comment (rafzi) [inline src/jitlibs/softfloat/CMakeLists.txt:366]: why was this required? a shallow search gave me the impression that this is the default already: https://stackoverflow.com/questions/5659225/how-do-i-set-the-options-for-cmake-ar

Comment (JoGei) [reply src/jitlibs/softfloat/CMakeLists.txt]: Good catch. I first modeled the build configuration after the original which used these specific archive commands. Not sure, whether these actually have any effect. On my setup, CMake defaults to `ar qc ...` not `ar crs ...` invoked by the original berkeley softfloat.

Removed with [commit a76b228](https://github.com/JoGei/etiss/commit/a76b22870118113702e31b61e755a697f2f4ca99)

File: JITImpl/GCC/GCCJIT.cpp [hunk 3]
```diff
@@ -136,24 +136,26 @@ void *GCCJIT::translate(std::string code, std::set<std::string> headerpaths, std
...
-            ss << "-L\"" << *iter << "\" ";
+            ss << " -L" << *iter << " ";
     }
+    
+    ss << "-o " << path_ << "lib" << codefilename << ".so " << path_ << codefilename << ".o";
+
+
     for (std::set<std::string>::const_iterator iter = libraries.begin();iter != libraries.end();iter++){
-            ss << "-l\"" << *iter << "\" ";
+            ss << " -l" << *iter << " ";
```
[the remark this answers is not in the document]
Comment (JoGei) [reply JITImpl/GCC/GCCJIT.cpp]: There is a special place somewhere for paths with spaces ;-)

Fixed with [commit 96d592d](https://github.com/tum-ei-eda/etiss/pull/18/commits/a76b22870118113702e31b61e755a697f2f4ca99)

File: src/jitlibs/softfloat/CMakeLists.txt (hunk 1 above)
Comment (JoGei) [reply src/jitlibs/softfloat/CMakeLists.txt:20]: Only for some Linux machines. 

Are you referring to [4. SoftFloat Package Directory Structure](http://www.jhauser.us/arithmetic/SoftFloat-3/doc/SoftFloat-source.html) as the documentation in question? I can make out the various possible configurations, but I am not sure how to configure this for all possible ETISS use(r) cases.

As far as I understand the ETISS main page, Linux x86-64 and Windows is supported/tested. This means the CMakeLists.txt should be able to detect and configure
- Linux x86-80 w/ 8086-SSE or 8086, or
- Windows and set Win64-MinGW-w64 w/ 8086-SSE or 8086?

Comment (rafzi) [reply src/jitlibs/softfloat/CMakeLists.txt:20]: As far as I understand this page https://cmake.org/cmake/help/latest/variable/CMAKE_CL_64.html , that variable is only set when using cl as compiler, which would not be the case when using mingw. The standard way to check architecture bitness is CMAKE_SIZEOF_VOID_P.

Comment (JoGei) [reply src/jitlibs/softfloat/CMakeLists.txt]: Introduced more specific host information retrieval with latest commits [61d12ed](https://github.com/tum-ei-eda/etiss/pull/18/commits/61d12ed7a0c54eb32a9ad8a58e913a2d80560657)

However, build on Windows host still fails with MSVC: 
grafik

It seems like, MSVC does not support the '__int128' type introduced through a common header all Windows targets of the source softfloat library. MinGW seems to support this natively.

I am no Windows expert, however, it seems like we will have to add a MSVC-target configuration to the berkeley softfloat library, or replace the header specifiying the '__int128' by a MSVC conform data type.

Right now, this pull request will break Windows support for ETISS.

**Comment by JoGei** [comment]
Comment on commit [da826fe](https://github.com/tum-ei-eda/etiss/pull/18/commits/da826feb5684bf8f995d875dc8eef56f362a9f87)
Build: Tested on Win10 with MSVC.
Simulation: Added random floating point operations to riscv example "helloworld". Build with -march=rv32gc.

| JIT | Win10 | Linux |
| ---- | --------- | ------- |
| TCC | pass | pass |
| GCC | NA | pass |
